cpp/src/reductions/segmented/compound.cuh

```diff
@@ -0,0 +1,193 @@
+  auto out_itr = result->mutable_view().template begin<ResultType>();
+
+  // Compute valid counts
+  rmm::device_uvector<size_type> valid_counts(num_segments, stream);
+  if (col.has_nulls() && (null_handling == null_policy::EXCLUDE)) {
+    auto valid_fn = [] __device__(auto p) -> size_type { return static_cast<size_type>(p.second); };
+    auto itr      = thrust::make_transform_iterator(d_col->pair_begin<InputType, true>(), valid_fn);
+    cudf::reduction::detail::segmented_reduce(itr,
+                                              offsets.begin(),
+                                              offsets.end(),
+                                              valid_counts.data(),
+                                              thrust::plus<size_type>{},
+                                              0,
+                                              stream);
+  } else {
+    thrust::adjacent_difference(
+      rmm::exec_policy(stream), offsets.begin() + 1, offsets.end(), valid_counts.begin());
+  }
+
+  // Run segmented reduction
+  if (col.has_nulls()) {
+    auto nrt = compound_op.template get_null_replacing_element_transformer<ResultType>();
+    auto itr = thrust::make_transform_iterator(d_col->pair_begin<InputType, true>(), nrt);
+    cudf::reduction::detail::segmented_reduce(
+      itr, offsets.begin(), offsets.end(), out_itr, compound_op, ddof, valid_counts.data(), stream);
+  } else {
+    auto et  = compound_op.template get_element_transformer<ResultType>();
+    auto itr = thrust::make_transform_iterator(d_col->begin<InputType>(), et);
+    cudf::reduction::detail::segmented_reduce(
+      itr, offsets.begin(), offsets.end(), out_itr, compound_op, ddof, valid_counts.data(), stream);
+  }
+
+  // Compute the output null mask
+  auto const bitmask                 = col.null_mask();
+  auto const first_bit_indices_begin = offsets.begin();
+  auto const first_bit_indices_end   = offsets.end() - 1;
+  auto const last_bit_indices_begin  = first_bit_indices_begin + 1;
+  auto [output_null_mask, output_null_count] =
+    cudf::detail::segmented_null_mask_reduction(bitmask,
+                                                first_bit_indices_begin,
+                                                first_bit_indices_end,
+                                                last_bit_indices_begin,
+                                                null_handling,
+                                                std::nullopt,
+                                                stream,
+                                                mr);
+  result->set_null_mask(std::move(output_null_mask), output_null_count);
+  return result;
+};
+
+template <typename ElementType, typename Op>
+struct segmented_result_dispatcher {
+ private:
+  template <typename ResultType>
+  static constexpr bool is_supported_v()
+  {
+    // the operator `mean`, `var`, `std` only accepts
+    // floating points as output dtype
```

Comment:
Can we write this explicitly, rather than relying on an assumption that all of our compound operators must return floating types? `is_supported_v` should also depend on `Op`.

Reply:
Seems adding explicit checks on the Op would be unnecessarily restrictive. Are you having an issue with the name of the functor? What if we changed the name to `compound_float_output_dispatcher` instead? This way it could be used by any future Op types without updating the `is_supported_v()` function.